Pattern Library: Prep strings for localization

chad1008 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to 5701-gh-Automattic/dotcom-forge

## Proposed Changes

- Mark all new strings for translation and removing the relevant `.eslintrc` file
- Specify unique `<title>` tag values for each pattern category
- Also updating the `H1` text to hopefully be more SEO friendly and keyword-relevant.
- Finally, updating the page titles slightly to avoid repetitively displaying "WordPress" or "WordPress.com" in those titles.

## Testing Instructions

- Confirm that each of the updated strings still appears as expected
- Test the `<title>` tags for the individual categories and make sure they are all properly populated
- Confirm that the `H1` and page `<title>` are displaying their new strings
- All checks should pass, specifically `ci/i18n` should report that there is sufficient translation for deployment

chad1008 committed:
prep copy-paste-info for l10n

chad1008 committed:
prep get-access-modal for l10n

chad1008 committed:
prep get-started for l10n

chad1008 committed:
prep header for l10n

chad1008 committed:
prep pattern-preview for l10n

chad1008 committed:
snake_case for consistency

chad1008 committed:
more SEO friendly H1 text

chad1008 committed:
update title tags

chad1008 commented on client/my-sites/patterns/components/pattern-library/index.tsx:
```diff
@@ -186,7 +186,7 @@ export const PatternLibrary = ( {
 			/>
 
-			<DocumentHead title={ translate_not_yet( 'WordPress Patterns - Category' ) } />
+			<DocumentHead title={ translate_not_yet( `WordPress ${ category } Patterns ` ) } />
```
I went with the previously recommended title here (see p1711382186987629-slack-C06ELHR6L9J), but I'm not sure if this format will always read the way we want, particularly after translation.

For example, "WordPress Form Patterns" reads well to me, but something like "WordPress About Patterns" feels a little off. Single quotes might help, i.e. "WordPress 'About' Patterns", but I'm not sure if we want it formatted that way or how that well that would work in terms of translation. Definitely open to input from reviewers 🤔

stephanethomas replied:
We can't really localize a string like so, i.e. where different pieces are concatenated and which includes a variable ([documentation](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-calypso/README.md#strings-only)). I believe we will have to follow [the approach](https://github.com/Automattic/wp-calypso/blob/975b3c2a995d3f39bd8e1904d1b90c0276133116/client/my-sites/themes/use-theme-showcase-logged-out-seo-content.js#L29-L35) of the `Themes` page where we define a different title for each category. We will have to come up with different meta descriptions for SEO purposes anyway (I don't know if we also want to have a different H1 for each category though).

chad1008 replied:
And now that I'm swapping in `translate`, I'm hitting the exactly the error you were referencing. There are a couple of other spots where we also have variables in the translation strings that i'll tackle at the same time.

Thanks for the point to the `Themes` page, that's a good starting point to work around this 🙏

chad1008 replied:
Done - I've intentionally structured the object holding the `title` translations so we can easily add `descriptions` later as well.

chad1008 committed:
the one that got away...

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -7,6 +8,7 @@ import ImgStyle from './images/style.svg';
+	const translate_not_yet = useTranslate();
 	return (
 		<PatternsSection
 			bodyFullWidth
```
We should also localize the `description` and `title` properties.

chad1008 replied:
Thanks for spotting that!

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -44,10 +53,13 @@ export function PatternsCopyPasteInfo() {
 							<img src={ ImgEdit } alt="" />
```
Tangentially related, should we also make sure that all images have an `alt` tag with a descriptive string for accessibility?

fredrikekelund replied:
I don't think these particular images convey relevant enough meaning to warrant `alt` texts.

stephanethomas replied:
I was curious, and had a look to our [accessibility checklist](https://github.com/Automattic/wp-calypso/blob/trunk/docs/accessibility-checklist.md#screen-reader-support) which mentions that all images should have an `alt` text. However, it also links to this [article](https://www.a11yproject.com/posts/alt-text/) which indeed recommends the following:

> Use an empty alt attribute for any image that is decorative or not necessary for understanding the content of the page. For example, an image providing decoration to an article's heading could be hidden by giving the image an alt="" or simply alt

stephanethomas reviewed (changes requested):
I think it would be preferable to revisit how we localize page titles. Apart from that, I've also noticed that the number of patterns is not localized ([client](https://github.com/Automattic/wp-calypso/blob/ecb0e3a369014ec24e745a3c1926f58dda6d1dd7/client/my-sites/patterns/components/category-gallery/client.tsx#L83-L88), [server](https://github.com/Automattic/wp-calypso/blob/d30b9682484316bcafac6cf91284cad8f0534aba/client/my-sites/patterns/components/category-gallery/server.tsx#L51-L56)). The rest is all good 😀.

stephanethomas commented:
Oh, and if you don't use the `Pre-merge Checklist` feel free to remove it from the description - that's less noise for the readers🙂.

chad1008 committed:
proper ellipsis for translation

chad1008 committed:
dynamic page titles for each category

chad1008 committed:
remove variables from `translate` args

chad1008 committed:
ready for translation

chad1008 commented on client/my-sites/patterns/components/category-not-found/index.tsx:
```diff
@@ -8,16 +8,27 @@ import './style.scss';
+	};
 	return (
 		<EmptyContent
-			title={ translate_not_yet( "Oops! We can't find this category!", { comment } ) }
-			line={ translate_not_yet( "The category you are looking for doesn't exist.", { comment } ) }
-			action={ translate_not_yet( 'Browse all patterns', { comment } ) }
```
the `comment` variable here was triggering a `wpcalypso/i18n-no-variables` warning, so I've replaced it with the actual string. The component props felt really messy with that repetition, so i've moved it (and the actual translation) into a separate variable.

I don't love that we have to repeat the same string over and over for the `comment`, but I don't see another way around that warning 🤔

chad1008 commented on client/my-sites/patterns/components/pattern-library/index.tsx:
```diff
@@ -176,6 +176,56 @@ export const PatternLibrary = ( {
 
 	const isHomePage = ! category && ! searchTerm;
 
+	const PATTERN_SEO_CONTENT: Record< Category[ 'name' ], { title: string } > = useMemo(
```
100% open to feedback on whether or not this is the best way to type this. TS was a little cranky at first and this is where I landed after experimenting a bit.

chad1008 committed:
translate `copy-paste-info` title and description

chad1008 committed:
translate pattern count string

chad1008 commented:
I _think_ I've covered all the bases from the last round of feedback. I left a couple of inline notes in the places I had to eliminate variables from `translate()` calls, just in case there's a better approach to solving those problems. 😄

> I've also noticed that the number of patterns is not localized ([client](https://github.com/Automattic/wp-calypso/blob/ecb0e3a369014ec24e745a3c1926f58dda6d1dd7/client/my-sites/patterns/components/category-gallery/client.tsx#L83-L88), [server](https://github.com/Automattic/wp-calypso/blob/d30b9682484316bcafac6cf91284cad8f0534aba/client/my-sites/patterns/components/category-gallery/server.tsx#L51-L56)). The rest is all good 😀.

Oh, you've got sharp eyes @stephanethomas, that was a sneaky one. single word tucked in there beneath the ternary. Should be good with 17f9acd124

fredrikekelund committed:
Translate singular/plural in CategoryGallery

fredrikekelund committed:
Apply changes from #88974

fredrikekelund committed:
Add comments

fredrikekelund committed:
Add more comments, no useMemo

fredrikekelund committed:
Fix type error

fredrikekelund reviewed (commented):
Thanks for taking care of this, @chad1008 👍

Overall, I think this is in pretty good shape, but I took the liberty of fixing a few small issues I found by pushing a few commits:
- Translate the full pattern count string (including the number) in `. patterns-category-gallery__item-count`
- Applied the changes from @wongasy's https://github.com/Automattic/wp-calypso/pull/88974
- Added translation comments

Regarding translation comments, I actually think we should keep adding even more of those, since they will help with translation quality. When working on Tumblr Domains, every new string we added also got a translation comment. 

Given that the `String Freeze` label is already applied, I think we should prioritize adding those comments to mitigate double work for translators.

cc @wongasy @stephanethomas

fredrikekelund commented:
I also took a quick stab at a "string repository" type approach in #88997. Again, I was inspired by what @nightnei started when we worked on Tumblr Domains. That solution seems to work as expected with our translation pipeline, which makes me think we might want to adopt it. 

We can land this PR first and that one later, though (if we decide to go that route).

fredrikekelund commented on client/my-sites/patterns/components/pattern-library/index.tsx:
```diff
@@ -176,6 +176,95 @@ export const PatternLibrary = ( {
 
 	const isHomePage = ! category && ! searchTerm;
 
+	const PATTERN_SEO_CONTENT: Record< Category[ 'name' ], { title: string } > = {
```
Me removing the `useMemo` call here might have been somewhat subjective, but I don't think these `translate` calls are expensive enough to make us gain much from `useMemo`. This article is something I regularly refer to when `useCallback` and `useMemo` comes up: https://kentcdodds.com/blog/usememo-and-usecallback

chad1008 replied:
Thanks @fredrikekelund! That's an article I find myself circling back to pretty regularly as well. I think I just mirrored `useMemo` from the `Themes` [implementation](https://github.com/Automattic/wp-calypso/blob/975b3c2a995d3f39bd8e1904d1b90c0276133116/client/my-sites/themes/use-theme-showcase-logged-out-seo-content.js#L29-L35) I was imitating and could have evaluated the need for it a bit more, but I agree with your point that it probably isn't needed here.

stephanethomas commented on client/my-sites/patterns/components/category-not-found/index.tsx:
```diff
@@ -8,16 +8,27 @@ import './style.scss';
-		'Message displayed when an invalid category was specified while searching block patterns';
+	const translate = useTranslate();
 
+	const emptyContentStrings = {
+		title: translate( "Oops! We can't find this category!", {
+			comment:
+				'Message displayed when an invalid category was specified while searching block patterns',
+		} ),
+		line: translate( "The category you are looking for doesn't exist.", {
+			comment:
+				'Message displayed when an invalid category was specified while searching block patterns',
+		} ),
+		action: translate( 'Browse all patterns', {
+			comment:
+				'Message displayed when an invalid category was specified while searching block patterns',
+		} ),
+	};
```
Since we must add those comments anyway, maybe we should make them a bit more specific?

```suggestion
	const emptyContentStrings = {
		title: translate( "Oops! We can't find this category!", {
			comment:
				'Heading displayed when an invalid category was specified while searching block patterns',
		} ),
		line: translate( "The category you are looking for doesn't exist.", {
			comment:
				'Message displayed when an invalid category was specified while searching block patterns',
		} ),
		action: translate( 'Browse all patterns', {
			comment:
				'Label of a button displayed when an invalid category was specified while searching block patterns',
		} ),
	};
```

stephanethomas replied:
By the way, here is another approach we could consider that gets rid of one level of indentation:

```javascript
const title = translate( "Oops! We can't find this category!", {
  comment:
    'Message displayed when an invalid category was specified while searching block patterns',
} );

const line = translate( "The category you are looking for doesn't exist.", {
  comment:
    'Message displayed when an invalid category was specified while searching block patterns',
} );

const action = translate( 'Browse all patterns', {
  comment:
    'Message displayed when an invalid category was specified while searching block patterns',
} );

return (
  <EmptyContent
    title={ title }
    line={ line }
    action={ action }
```

chad1008 replied:
I like both of these suggestions... the more specific comments but also the individual variables in favor of the single object. Thanks!

stephanethomas commented:
> Regarding translation comments, I actually think we should keep adding even more of those, since they will help with translation quality. When working on Tumblr Domains, every new string we added also got a translation comment.

I'm not sure we need a comment for each and every string, and our conventions for that matter don't enforce that ([Core](https://codex.wordpress.org/I18n_for_WordPress_Developers#Descriptions), [Calypso](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-calypso/README.md#options)). Comments are especially useful to explain dates, placeholders, or references. A good example from this pull request is the following string:

```javascript
{ translate( 'Make it yours' ) }
```

Since they don't have any context (what does `it` refers to?), translators won't be able to provide a good translation for that string. We should add a comment that provides some clue, e.g.:

> This string refers to block patterns, and their ability to be customized to build a site.

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -44,10 +55,13 @@ export function PatternsCopyPasteInfo() {
 
-						<div className="section-patterns-info__item-title">Make it yours</div>
+						<div className="section-patterns-info__item-title">
+							{ translate( 'Make it yours' ) }
```
We should add a translator comment, as mentioned [here](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

fredrikekelund commented:
> I'm not sure we need a comment for each and every string

I agree we don't need it for every single string. If we're on the fence, though, then maybe one comment too many is still preferable to one too few.

> A good example from this pull request

Agreed 👍

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -7,12 +8,15 @@ import ImgStyle from './images/style.svg';
+			) }
 			theme="dark"
-			title="Copy, paste, customize—it’s easy like that"
+			title={ translate( 'Copy, paste, customize—it’s easy like that' ) }
```
We should probably add a comment here to provide translators with [context](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -21,9 +25,13 @@ export function PatternsCopyPasteInfo() {
 
-						<div className="section-patterns-info__item-title">Copy-paste your way</div>
+						<div className="section-patterns-info__item-title">
+							{ translate( 'Copy-paste your way' ) }
```
We should probably add a comment here to provide translators with [context](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -32,10 +40,13 @@ export function PatternsCopyPasteInfo() {
 
-						<div className="section-patterns-info__item-title">Bring your style with you</div>
+						<div className="section-patterns-info__item-title">
+							{ translate( 'Bring your style with you' ) }
```
We should probably add a comment here to provide translators with [context](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

stephanethomas commented on client/my-sites/patterns/components/copy-paste-info/index.tsx:
```diff
@@ -56,10 +70,13 @@ export function PatternsCopyPasteInfo() {
 
-						<div className="section-patterns-info__item-title">Responsive by design</div>
+						<div className="section-patterns-info__item-title">
+							{ translate( 'Responsive by design' ) }
```
We should probably add a comment here to provide translators with [context](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

stephanethomas commented on client/my-sites/patterns/components/get-access-modal/index.tsx:
```diff
@@ -43,25 +45,28 @@ export const PatternsGetAccessModal = ( {
 				<div className="patterns-get-access-modal__inner">
-					<div className="patterns-get-access-modal__title">Unlock the full pattern library</div>
+					<div className="patterns-get-access-modal__title">
+						{ translate( 'Unlock the full pattern library' ) }
```
We should probably add a comment here to provide translators with [context](https://github.com/Automattic/wp-calypso/pull/88871#issuecomment-2025218632).

stephanethomas commented on client/my-sites/patterns/components/get-started/index.tsx:
```diff
@@ -35,8 +37,10 @@ export function PatternsGetStarted() {
-						<div className="patterns-get-started__item-description">Block Patterns</div>
+						<div className="patterns-get-started__item-name">{ translate( 'Video tutorial' ) }</div>
+						<div className="patterns-get-started__item-description">
+							{ translate( 'Block Patterns' ) }
```
I wonder if we shouldn't add a comment for translators here as well.

stephanethomas commented on client/my-sites/patterns/components/pattern-library/index.tsx:
```diff
@@ -176,6 +176,95 @@ export const PatternLibrary = ( {
+				textOnly: true,
+			} ),
+		},
+	};
```
We could probably define this array outside of the component (i.e. just above) so the latter is easier to skim.

fredrikekelund replied:
Either that, or we move this, and `<DocumentHead title={ seoContent.title } />` into its own component.

Moving this array outside the function body is slightly complex because we need to pass the `translate` function for that to work (we could theoretically import `translate` directly, but `useTranslate` is the recommended approach). #88997 could serve as inspiration for how to do that, though.

chad1008 replied:
Holding off on changes here, as I know there's still ongoing disucssion over in #88997, but I'm happy to pursue whichever course y'all think best!

fredrikekelund replied:
Let's go with adding a new component that wraps `DocumentHead`. Within that component, we can define `PATTERN_SEO_CONTENT` in the function scope, no problem

stephanethomas commented on client/my-sites/patterns/components/pattern-preview/index.tsx:
```diff
@@ -171,7 +198,11 @@ function PatternPreviewFragment( {
+						<Icon height={ 18 } icon={ lock } width={ 18 } />{ ' ' }
+						{ translate( 'Get access', {
+							comment:
+								'Button label for that shows in Pattern Library instead of "Copy" when logged-out users need to sign up',
```
I would avoid the mention to the other label, and instead explain what kind of feature the user will get access to:
```suggestion
								'Button label shown when logged-out users need to sign up to be able to use a pattern',
```

fredrikekelund replied:
Good call 👍

stephanethomas commented:
> If we're on the fence, though, then maybe one comment too many is still preferable to one too few.

Definitively 👍.

stephanethomas reviewed (approved):
My feedback is mostly about adding comments for some strings that may be to abstract for translators. There are merge conflicts to fix but other than that, we're good!

stephanethomas commented:
Oh by the way, I've found something interesting with search results:

| `/patterns/header?grid=1&s=image` | `/patterns/header?grid=1&s=cool`
| -----  | -----
| ![image](https://github.com/Automattic/wp-calypso/assets/594356/4affed44-d518-4feb-86c1-cc0c76b82962) | ![image](https://github.com/Automattic/wp-calypso/assets/594356/71540bd1-ccb3-4af0-89a1-f3f2f3eb62bc)

The plural form is indeed displayed when the number of patterns is zero. I don't think that is caused by your changes but I thought I would mention it.

chad1008 committed:
clean up category not found strings

chad1008 committed:
Update client/my-sites/patterns/components/pattern-preview/index.tsx

Co-authored-by: Stéphane Thomas <email@stephanethomas.com>

chad1008 committed:
add translator comments

fredrikekelund committed:
More translation comments

fredrikekelund committed:
One more comment

fredrikekelund commented on client/my-sites/patterns/components/pattern-library/index.tsx:
```diff
@@ -189,18 +278,18 @@ export const PatternLibrary = ( {
 					setSearchTerm( query );
 				} }
-				title={ translate_not_yet( 'It’s Easier With Patterns' ) }
+				title={ translate( 'Build your site faster with patterns' ) }
```
```suggestion
				title={ translate( 'It's Easier With Patterns' ) }
```
Was this changed based on this Slack thread? p1711383014975249-slack-C06ELHR6L9J

The existing copy has been written by the marketing team and was updated in #88622. Let's stick with that until we have some other consensus.

fredrikekelund replied:
Please take a look at this before merging, though, @chad1008

chad1008 replied:
Yep, that's exactly where that came from. I've committed your suggestion to leave it as it was for now, thank you for the guidance!

fredrikekelund committed:
Even more comments

fredrikekelund reviewed (approved):
I've gone over this one more time and added a few more comments, so I think this should be good to merge now 👍 I'm unsure if you missed pushing some changes, @chad1008, since a couple of @stephanethomas's comments were marked as resolved even if no comments were added. No big deal, but just be mindful of any merge conflicts that may come up with my commits.

chad1008 committed:
Update client/my-sites/patterns/components/pattern-library/index.tsx

Co-authored-by: Fredrik Rombach Ekelund <fredrik@fredrik.computer>

chad1008 committed:
fix broken quotes from last suggestion

chad1008 committed:
move `DocumentHead` call and SEO content into isolated component

chad1008 commented:
I've addressed the remaining feedback, including moving the abstracting our SEO strings into a separate component wrapping `DocumentHead`. With CI checks passing I'm going to proceed with the merge, but if we want to adjust anything on that new wrapped component I'm happy to address it on Monday (cc @fredrikekelund )
